interpreter/cling/lib/Interpreter/NullDerefProtectionTransformer.cpp

```diff
@@ -49,29 +50,33 @@ namespace cling {
     PointerCheckInjector(Sema& S) : m_Sema(S), m_Context(S.getASTContext()),
     m_clingthrowIfInvalidPointerCache(0) {}
 
-    void VisitStmt(Stmt* S) {
+    bool VisitStmt(Stmt* S) {
       for (auto child: S->children()) {
         if (child)
-          Visit(child);
+            VisitStmt(child);
       }
+      return true;
     }
 
-    void VisitUnaryOperator(UnaryOperator* UnOp) {
-      Visit(UnOp->getSubExpr());
-      if (UnOp->getOpcode() == UO_Deref) {
-        UnOp->setSubExpr(SynthesizeCheck(UnOp->getSubExpr()));
-      }
+    bool VisitUnaryOperator(UnaryOperator* UnOp) {
+      VisitStmt(UnOp->getSubExpr());
+      if (UnOp->getOpcode() == UO_Deref)
+        if (UnOp->getSubExpr()->getType().getTypePtr()->isPointerType())
+          UnOp->setSubExpr(SynthesizeCheck(UnOp->getSubExpr()));
+      return true;
     }
 
-    void VisitMemberExpr(MemberExpr* ME) {
-      Visit(ME->getBase());
-      if (ME->isArrow()) {
-        ME->setBase(SynthesizeCheck(ME->getBase()));
+    bool VisitMemberExpr(MemberExpr* ME) {
+      if (!((VarDecl*)(ME->getMemberDecl()))->isStaticDataMember()) {
```

Comment:
I understand that we don't want to synthesize for `MemberExpr` on static members. But why would we not want to recurse? I.e. I'd have expected this to be `if (ME->isArrow() && the_member_is_no_static)` below the `VisitStmt()`.